### DISCUSSION TITLE
Building custom firmware with different UART PINs, how the uart flash process work in this case?

### DISCUSSION THREAD

**Discussion by hitech95** [darkxst/silabs-firmware-builder#26] [Q&A]
Hi,

I need to flash a module and change the UART pins (both TX RX and HW_CTRL ones) as the module is connected using other pins.
To do so I thought to create a custom configuration and build a new FW to then upload it via the WebUI.

The question is about what pins to use for the Upload, do I have to use the generic ones or
 the ones used by the current installed Firmware?

**Comment by darkxst** [comment]
You just need to patch the pins for your firmware.

When you upload with flasher your just connecting to whatever is set in the bootloader and firmware.

**Reply by hitech95** [reply]
Is the firmware builder also building/flashing the bootloader?

Lets say that right now my module is connected to PD00(TX) and PD03(RX).
Is there a way to verify the pins used by the bootloader?

**Reply by darkxst** [reply]
> Is the firmware builder also building/flashing the bootloader?

No its not including bootloader.

I dont know any way to verify pins used by bootloader but it should be the same as firmware (except doesnt use flow control)

**Reply by hitech95** [reply]
Ok, my last question I have not cheked the code yet but I suppose that the web page uses an uart command to jump to booloader and then it proceed with the update. Is this correct?

So lets say if the bootloader uses PB00 and PB01 I always have to use that pins to upgrade the firmware even if the firmware might be built for other pins.

Now the missing questions is how to calculate the CTUNE and if its necessary!

**Reply by darkxst** [reply]
Universal-silabs-flasher send special reset codes of serial to enter bootloader (these vary across the three firmware types)

the pins need to match physical connection on chip. Bootloader and firmware should both use same settings, if you were building a custom device you can set whatever pins you like when building bootloader 

CTUNE depends on the PCB design, but most are designed with at the default 140, or 128

**Comment by darkxst** [comment]
what device is this for anyway?

**Reply by hitech95** [reply]
Its a Zyxel router with the zigbee chip in it.
I'm still trying to figure out how the hell it works :D
